log/mock.go

```diff
@@ -108,3 +116,39 @@ func (m *Mock) Clear() {
 	w := m.w.(*mockWriter)
 	w.clearChan <- struct{}{}
 }
+
+type waitingMockWriter struct {
+	logChan chan string
+}
+
+// newWaitingMockWriter returns a new waitingMockWriter
+func newWaitingMockWriter() *waitingMockWriter {
+	logChan := make(chan string)
+	return &waitingMockWriter{
+		logChan,
+	}
+}
+
+func (m *waitingMockWriter) logAtLevel(p syslog.Priority, msg string) {
+	m.logChan <- fmt.Sprintf("%s: %s", levelName[p&7], msg)
+}
+
+// WaitForMatch returns the first log line matching a regex. It accepts a
+// regexp string and timeout. If the timeout value is met before the
+// matching pattern is read from the channel, an error is returned.
+func (m *Mock) WaitForMatch(reString string, timeout time.Duration) (string, error) {
+	w := m.w.(*waitingMockWriter)
```

Comment:
Might want to use the `w, ok := ...` version of the type assertion here, and fail if the type assertion fails.

Reply:
Ooh this is a good catch. This type assertion is actually a little unusual. We would normally expect the receiver to be `m *waitingMockWriter`, but we can't do that here because `WaitForMatch` needs to be an exported method and `waitingMockWriter` is not exported. Well, we might actually be able to get away with it from a compiler perspective but it would be awkward.

The existing GetAllMatching got away with this because we knew that `Mock` would always be insantiated with `mockWriter`. Now that we have two possibilities we've introduced the possibility that you could do `m := NewWaitingMock()` and then call `m.GetAllMatching(...)`, or `m := NewMock()` and then call `m.WaitForMatch(...)`. Either of those calls would panic.

In general that's something we strongly try to avoid in production code. Given that this is test code, I'm okay with it so long as we document it. We could also potentially figure out ways to solve this more cleanly. I'm not coming up with ideas at the moment but I'm sure they exist. :-D

Reply:
(that said, I do approve of returning an error rather than panicking, even for test code)

Reply:
Thanks! I have made a new WaitingMock struct embedding impl and moved the methods to that. This avoids the type assertion problems and hopefully confusion for the user of the mocks about what methods are available.